ada/language/ast.py

```diff
@@ -36,6 +37,44 @@ def default_origin():
     return (origin, No(T.AdaNode))
 
 
+def default_imprecise_fallback():
+    """
+    Helper to return an imprecise fallback dynamic param spec which defaults to
+    False.
+    """
+    return (imprecise_fallback, False)
+
+
+def _referenced_decl_internal_impl(subject, try_immediate):
+    """
+    Helper to generate the piece of logic that is common to all
+    "referenced_decl_internal" implementations. "subject" is the element which
+    holds the ref_var.
+    """
+    compute_logic_val = Self.logic_val(
+        Entity, subject.ref_var, try_immediate
+    )
+
+    try_compute_logic_val = Try(
+        compute_logic_val,
+        LogicValResult.new(success=False, value=No(AdaNode.entity))
+    )
+
+    precise_body = compute_logic_val.value.cast_or_raise(T.BasicDecl.entity)
+
+    imprecise_body = Let(lambda v=try_compute_logic_val: Let(
```

Comment:
With that code  you'll get the imprecise_body even the user didn't ask for the imprecise fallback

Reply:
I don't think so, the Let only appears in the "then" branch of the If expression. I checked the generated code and it is as I expect? I would agree with you if this was a `Var` since those are a syntactic sugar for a top-level `Let`, but it's not the case here.

Reply:
Ah, OMG that's true, that's it you know the DSL better than I do!

I'm not sure I like this style though. Very much like the other comment I made, that's abusing the implicit macro system we have in the DSL. First, most of the time an expression assigned to a variable that is not in a Var() is an error. Second, that won't transition well to the concrete syntax - the code will just be inlined down, so might as well do it already :)